dynamic/json_test.go

```diff
@@ -87,8 +88,43 @@ func TestMarshalJSONIndent(t *testing.T) {
 	testutil.Eq(t, "{\n\t\"foo\": [\n\t\t\"VALUE1\"\n\t],\n\t\"bar\": \"bedazzle\"\n}", string(jsIndent))
 }
 
+func TestUnmarshalJSON(t *testing.T) {
+	md, err := desc.LoadMessageDescriptorForMessage((*testprotos.TestRequest)(nil))
+	testutil.Ok(t, err)
+
+	js1 := []byte(`{"foo":["VALUE1"],"bar":"bedazzle"}`)
+	dm1 := NewMessage(md)
+	err = dm1.UnmarshalJSON(js1)
+	testutil.Ok(t, err)
+	foo1 := dm1.GetFieldByNumber(1)
+	bar1 := dm1.GetFieldByNumber(2)
+	testutil.Eq(t, coerceSlice([]int32{1}, reflect.Int32), foo1)
+	testutil.Eq(t, coerceSlice("bedazzle", reflect.String), bar1)
+
+	js2 := []byte("{\n  \"foo\": [\n    \"VALUE1\"\n  ],\n  \"bar\": \"bedazzle\"\n}")
+	dm2 := NewMessage(md)
+	err = dm2.UnmarshalJSON(js2)
+	testutil.Ok(t, err)
+	foo2 := dm2.GetFieldByNumber(1)
+	bar2 := dm2.GetFieldByNumber(2)
+	testutil.Eq(t, coerceSlice([]int32{1}, reflect.Int32), foo2)
+	testutil.Eq(t, coerceSlice("bedazzle", reflect.String), bar2)
+}
+
 func TestUnmarshalJSONAllowUnknownFields(t *testing.T) {
-	// TODO
+	md, err := desc.LoadMessageDescriptorForMessage((*testprotos.TestRequest)(nil))
+	testutil.Ok(t, err)
+	js := []byte(`{"foo":["VALUE1"],"bar":"bedazzle","xxx": 1}`)
+	dm := NewMessage(md)
+	err = dm.UnmarshalJSON(js)
+	testutil.Nok(t, err)
+	unmarshaler := &jsonpb.Unmarshaler{AllowUnknownFields: true}
+	err = dm.UnmarshalJSONPB(unmarshaler, js)
+	testutil.Ok(t, err)
+	foo := dm.GetFieldByNumber(1)
+	bar := dm.GetFieldByNumber(2)
+	testutil.Eq(t, coerceSlice([]int32{1}, reflect.Int32), foo)
```

Comment:
This call to `coerceSlice` is not necessary since elements of `[]int32{1}` are already `reflect.Int32` type.

This is used for testing dynamic message because it allows numeric promotion -- like from an int32 to an int64. So I used it in a generic test case where inputs are parameterized (and sometimes input may be a slice of int32 and others a slice of int64, etc). So it was a utility function to convert everything to int64 to produce the "expected" value -- because you can set an int64 field with an int32 value, but querying the value back always returns int64s.

Hope that makes sense, but here you can do the simple test:

```go
testutil.Eq(t, []int32{1}, foo)
```